pyccel/ast/core.py

```diff
@@ -179,85 +191,106 @@ def __init__(self, var):
 
 # Pow, Add, Mul need to inherite sympy.Boolean to be able to use them in a logical expression
 
-class Pow(sp_Pow, sp_Boolean, PyccelAstNode):
-    def __new__(cls, *args, evaluate = False, **kwargs):
-        return sp_Pow.__new__(cls, *args, evaluate = evaluate, **kwargs)
-
-    def __init__(self, *args, evaluate = False, **kwargs):
+class PyccelOperator(Expr, PyccelAstNode):
+    @property
+    def rank(self):
         # TODO: Use broadcasting rules to decide shape (https://docs.scipy.org/doc/numpy/user/basics.broadcasting.html)
-        self._rank = max(getattr(a,'rank',0) for a in self._args)
+        return max(getattr(a,'rank',0) for a in self._args)
```

Comment:
As discussed previously. Did we not want to move this to the constructor and the use the PyccelAstNode definition of def rank? The rank will not change and this prevents it being recalculated multiple times

Reply:
yes we should but  it was not  previously on the constructor it was just a property,
so i left it as it is, I didn't want to implement it and preferred to leave it for another pull request

Reply:
I don't see what you mean. The original code for `Pow` was setting `self._rank` in the `__init__`, but now you removed that. If you do not need an `__init__` in the subclasses of `PyccelOperator`, this one will be called automatically.

Reply:
the diff here is wrong  you can verify the `Add`, `Mul` methods this what was changes the `Pow`
was completely removed and not just changed

Reply:
Is there a reason for not writing the following?
```
class PyccelOperator(Expr, PyccelAstNode):

    def __init__(self, *args, **kwargs):
        self._rank = max(getattr(a,'rank',0) for a in self._args)
```
This should work and it would be called automatically by any subclass that does not reimplement an `__init__` method. (I don't think `Expr` has an `__init__` method, so there should be no ambiguity on which constructor gets called.)

Reply:
yes we can , but I prefer to keep the attribute the None for the moment otherwise we will have to add them to all the nodes and modify the type inference mechanism to take that into account,
it's best to do it separately in another pull request

Reply:
OK. Do we have a related issue where we can write this down?

Reply:
I don't understand. Why does already having a rank for these objects mess up the type inference mechanism?

Reply:
it doesn't mess up anything , just wanted to do it separately

Reply:
yes @yguclu  we have this
https://github.com/pyccel/pyccel/issues/305

Reply:
This issue does not mention initialising values in the constructor rather than at each function call. Do you want to add that stipulation to the issue?
Personally I don't see how the 2 are related. Calling X.rank already gives a result at the moment even if _rank is still None. I'm worried that if we don't move this code while we remember it then we will forget as an error cannot appear due to this

Reply:
it does mention that , `we may need to use the PyccelAstNode attributes where we infer the type automatically at the instance creation`.
the thing is those attributes where never used before so adding them shouldn't give any error because they were not used only in the basic cases like the `Variable` class

Reply:
May and should are not the same thing ^^
I know they won't give an error. I meant that when we start using the attribute consistently we might use errors to track down the places where the rank has still not been initialised. In that case this wouldn't show up.
Given that it is a 1 line commit I would have thought it would have been easier to do it here the first time it's used rather than add the bad practice code to master and risk missing it when you come back to it. But if you prefer to do it later then ok. I have added a comment to the issue to clarify that we need to ensure initialisation is carried out in `__init__` where possible